index.html

```diff
@@ -18852,4 +18929,495 @@
+		 */
+		compareSingle(currentMessage, historyMessage) {
+			return this.keywordSimilarities(currentMessage, historyMessage)
+		}
+
+		/**
+		 * Compares a single message across the entire history (or memories)
+		 */
+		compareSeries(currentMessage, ...history) {
+			let comparisons = history.map(historyMessage => {
+				return { snippet: historyMessage, match: this.compareSingle(currentMessage, historyMessage) }
+			});
+			comparisons.sort((a, b) => {
+				return a.match > b.match ? -1 : 1
+			});
+			return comparisons;
+		}
+
+		/**
+		 * Compares a single message across the entire history, with the amount of values to return (like a top N) and a minimum similarity as a percentage
+		 */
+		compareAndGetTopNWithCull(currentMessage, valuesToReturn, similarity, ...history) {
+			let comparisonsSorted = this.compareSeries(currentMessage, ...history);
+			if (!!similarity) {
+				let comparisonsToReturn = []
+				for (let i = 0; i < comparisonsSorted.length; i++) {
+					if (comparisonsSorted[i].match < similarity) {
+						break;
+					}
+					comparisonsToReturn.push(comparisonsSorted[i])
+				}
+				comparisonsSorted = comparisonsToReturn;
+			}
+			return comparisonsSorted.filter((elem, pos, arr) => arr.findIndex(elemIn => elemIn.snippet === elem.snippet) === pos).slice(0, valuesToReturn);
+		}
+
+		/**
+		 * Compares a single message across the entire history, with the amount of values to return (like a top N)
+		 */
+		compareAndGetTopN(currentMessage, valuesToReturn, ...history) {
+			return this.compareAndGetTopNWithCull(currentMessage, valuesToReturn, undefined, ...history);
+		}
+
+		/**
+		 * Should not be used - only for internal use with the function getParagraphsFromHistory
+		 */
+		_currentContextCache
+		_gameTextCache
+		_lastResponseCache
+		_paragraphsCache
+
+		/**
+		 * Gets a static copy of the entire game text
+		 */
+		getAllGameText() {
+			if (this._gameTextCache === undefined) {
+				let gameText = concat_gametext().replace(/\xA0/g, ' ').trim()
+				let maxAllowedCharacters = getMaxAllowedCharacters(gameText, this._amountToTrimForContext);
+				if (gameText.length > maxAllowedCharacters) {
+					this._currentContextCache = gameText.substring(gameText.length - maxAllowedCharacters)
+					this._gameTextCache = gameText.substring(0, gameText.length - maxAllowedCharacters)
+				}
+				else {
+					console.log("History too short to enable memory fragments")
+					this._gameTextCache = ""
+					this._currentContextCache = ""
+				}
+			}
+			return this._gameTextCache;
+		}
+
+		/**
+		 * Gets the current context
+		 */
+		getCurrentContext() {
+			this.getAllGameText()
+			return this._currentContextCache
+		}
+
+		/**
+		 * Escapes a string to a regex supported format
+		 */
+		_escapeRegExp(stringToEscape) {
+			return stringToEscape.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
+		}
+
+		/**
+		 * Gets the last response from the AI and yourself
+		 */
+		getLastResponse() {
```

Comment:
Maybe add here to remove all specified characters like punctuation marks?  Not sure if this is the right portion...

```this._lastResponseCache = this._lastResponseCache.replace(/[.,\/#!$%\^&\*;:{}=\-_`~()\[\]]/g, "");```

Reply:
I do not think this is needed.  Different searchers in future may require those characters - so I would say leave them.  This is negated by the Minisearch lib anyway, as it splits by words and gets rid of special characters anyway as far as I understand it.